### PR TITLE
feat(agent): Abstract file storage access

### PR DESCRIPTION
**Pull request by kcze**
This improvement abstracts file access across whole project, so that AutoGPT app/agents can use different backends (e.g. local, GCS, S3). Changes:

- Rename *FileWorkspace* classes to *FileStorage* and add `exists`, `make_dir`, `delete_dir`, `list_folders` methods to all backends
- Rename `WORKSPACE_BACKED` env var to `FILE_STORAGE_BACKEND`
- Add `AgentFileManagerMixin` that replaces `AgentFileManager`
- Updates across codebase to adjust for this change, such as replacing direct file access to use `FileStorage` instead
- Updates to tests

Fixes related issue: OPEN-103

### CODE REVIEW

File: autogpts/autogpt/agbenchmark_config/benchmarks.py [hunk 1]
```diff
@@ -55,13 +58,20 @@ def bootstrap_agent(task: str, continuous_mode: bool) -> Agent:
         history=Agent.default_settings.history.copy(deep=True),
     )
 
+    local = config.file_storage_backend == FileStorageBackendName.LOCAL
+    restrict_to_root = not (local and not config.restrict_to_workspace)
```
Comment (Pwuts) [inline autogpts/autogpt/agbenchmark_config/benchmarks.py]: equivalent:
```suggestion
    restrict_to_root = not local or config.restrict_to_workspace
```

File: autogpts/autogpt/autogpt/agent_manager/agent_manager.py [hunk 2]
```diff
@@ -4,45 +4,46 @@
...
+        """Load the state of the agent with the given ID."""
+        state_file_path = Path(agent_id) / "state.json"
+        if not self.file_manager.exists(state_file_path):
             raise FileNotFoundError(f"Agent with ID '{agent_id}' has no state.json")
 
-        state = AgentSettings.load_from_json_file(state_file)
-        state.agent_data_dir = agent_dir
-        return state
+        text = self.file_manager.read_file(state_file_path)
+        return AgentSettings.parse_raw(text)
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/agent_manager/agent_manager.py]: `text`?

File: autogpts/autogpt/autogpt/agents/features/agent_file_manager.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from ..base import BaseAgent
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/agents/features/agent_file_manager.py]: Please move this import to the bottom of the import block. Now it's splitting core imports `typing` and `logging`.

File: autogpts/autogpt/autogpt/file_storage/base.py [hunk 4]
```diff
@@ -1,49 +1,55 @@
...
 from autogpt.core.configuration.schema import SystemConfiguration
 
 logger = logging.getLogger(__name__)
 
 
-class FileWorkspaceConfiguration(SystemConfiguration):
+class FileStorageConfiguration(SystemConfiguration):
     restrict_to_root: bool = True
-    root: Path = Path("/")
+    root: Path = Path(".")
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/file_storage/base.py]: What is the reason for this change?

File: autogpts/autogpt/autogpt/file_storage/gcs.py [hunk 5]
```diff
@@ -14,21 +14,21 @@
...
-from .base import FileWorkspace, FileWorkspaceConfiguration
+from .base import FileStorage, FileStorageConfiguration
 
 logger = logging.getLogger(__name__)
 
 
-class GCSFileWorkspaceConfiguration(FileWorkspaceConfiguration):
-    bucket: str = UserConfigurable("autogpt", from_env="WORKSPACE_STORAGE_BUCKET")
+class GCSFileStorageConfiguration(FileStorageConfiguration):
+    bucket: str = UserConfigurable("autogpt", from_env="STORAGE_BUCKET")
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/file_storage/gcs.py]: `WORKSPACE_STORAGE_*` was introduced in the last release, so we might want to still read those env variables for backwards compatibility. Unless nobody is using this functionality (gotta check Sentry).

Comment (Pwuts) [reply autogpts/autogpt/autogpt/file_storage/gcs.py]: Update: Sentry makes it look like we're the only ones using non-local storage ([link](https://significant-gravitas.sentry.io/discover/homepage/?field=title&field=event.type&field=project&field=timestamp&field=environment&field=release&field=user&field=server_name&id=26414&name=&query=stack.module%3Aautogpt.file_workspace.gcs+OR+stack.module%3Aautogpt.file_workspace.s3&sort=-timestamp&statsPeriod=30d&topEvents=5&yAxis=count%28%29)).

Why we can't be sure:
* Transactions are only recorded for users that opt-in and use the `AgentProtocolServer`
* Errors are only recorded for users that opt-in
  * As long as the implementations work properly (which they do), no errors will occur and nothing will be recorded in Sentry

File: autogpts/autogpt/autogpt/file_storage/base.py [hunk 6]
```diff
@@ -128,37 +148,36 @@ def _sanitize_path(
...
-                f"in workspace '{root}'."
+                f"in storage '{self.root}'"
             )
 
-        full_path = root.joinpath(relative_path).resolve()
+        full_path = self.root / relative_path
+        if self.is_local:
+            full_path = self.root.joinpath(relative_path).resolve()
+
+        full_path = Path(os.path.normpath(full_path))
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/file_storage/base.py]:
```suggestion
        full_path = self.root / relative_path
        if self.is_local:
            full_path = full_path.resolve()
        else:
            full_path = Path(os.path.normpath(full_path))
```

File: autogpts/autogpt/autogpt/file_storage/gcs.py [hunk 7]
```diff
@@ -103,11 +108,50 @@ def list(self, path: str | Path = ".") -> list[Path]:
...
+    def delete_dir(self, path: str | Path) -> None:
+        """Delete an empty folder in the storage."""
+        # Since GCS does not have directories, we don't need to do anything
+        pass
+
+    def exists(self, path: str | Path) -> bool:
+        """Check if a file or folder exists in GCS storage."""
+        path = self.get_path(
+            path
+        )  # Assuming this method returns a Path object or a string path
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/file_storage/gcs.py]: No need to assume, that's exactly what it does according to the declaration of `FileStorage.get_path`.
```suggestion
        path = self.get_path(path)
```

File: autogpts/autogpt/tests/utils.py [hunk 8]
```diff
@@ -10,5 +10,5 @@ def skip_in_ci(test_function):
     )(test_function)
 
 
-def get_workspace_file_path(workspace, file_name):
-    return str(workspace.get_path(file_name))
+def get_workspace_file_path(storage, file_name):
+    return str(storage.get_path(file_name))
```
Comment (Pwuts) [inline autogpts/autogpt/tests/utils.py]: This function isn't used anywhere, may be removed

File: autogpts/autogpt/autogpt/app/main.py [hunk 9]
```diff
@@ -304,14 +317,14 @@ async def run_auto_gpt(
             or agent_id
         )
         if save_as_id and save_as_id != agent_id:
-            agent.set_id(
-                new_id=save_as_id,
-                new_agent_dir=agent_manager.get_agent_dir(save_as_id),
-            )
+            # HACK this shouldn't be done this way
+            # preferably agents should not change their ID
+            # this just replaces the agent file space with a new one
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/app/main.py]: How would you propose organizing this properly?

File: autogpts/autogpt/autogpt/agents/utils/file_manager.py
Comment (Pwuts) [inline autogpts/autogpt/autogpt/agents/utils/file_manager.py:1]: Architecturally this is a pretty clean implementation, but this way the `FileManager` is nothing more than a proxy for the `FileStorage`, requiring a lot of duplicate code and juggling with root paths. Can we eliminate `FileManager` by using `FileStorage` directly, with the addition of a `FileStorage.clone_with_subroot` method?

File: autogpts/autogpt/autogpt/agents/features/agent_file_manager.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+    async def log_operation(self, content: str) -> None:
+        """Log an operation to the agent's log file."""
+        logger.debug(f"Logging operation: {content}")
+        self._logs_cache.append(content)
+        await self.files.write_file(self.LOGS_FILE, "\n".join(self._logs_cache) + "\n")
+
+    def get_logs(self) -> list[str]:
+        """Get the agent's logs."""
+        return self._logs_cache
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/agents/features/agent_file_manager.py]: - This is somewhat confusing use of term `logs`, synonymous with "log lines"
- General naming ("log") for something that has a very specific purpose (logging file operations)

Meta: the file operations logger is a non-ideal solution to a teething problem. However, with the current state of the codebase, we may be able to eliminate the file operations logger entirely by getting historical info from the `agent.event_history`.

File: autogpts/autogpt/autogpt/app/main.py [hunk 11]
```diff
@@ -304,14 +317,14 @@ async def run_auto_gpt(
...
+            agent.files = FileManager(agent._file_storage, f"agents/{save_as_id}/")
             # TODO: clone workspace if user wants that
             # TODO: ... OR allow many-to-one relations of agents and workspaces
 
-        agent.state.save_to_json_file(agent.file_manager.state_file_path)
+        await agent.save_state()
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/app/main.py]: ⚠️ When the `{save_as_id}` agent is loaded, its `self.files` and `self.workspace` will still point to the old agent's files and workspace, because those paths are [derived from the `agent_id`](https://github.com/Significant-Gravitas/AutoGPT/blob/e4a6dea87c5e4a1fb79529064b106d02029d4b2a/autogpts/autogpt/autogpt/agents/features/agent_file_manager.py#L54-L55).

**Review by Pwuts — CHANGES_REQUESTED** [review]
For a first XL pull request this looks very good! I do have a number of comments/requests but don't let that be a discouragement. :)

Questions:
* Have the added methods on `GCSFileStorage` and `S3FileStorage` been tested?
* How much of this has been end-to-end tested?

File: autogpts/autogpt/autogpt/file_storage/base.py (hunk 4 above)
Comment (kcze) [reply autogpts/autogpt/autogpt/file_storage/base.py]: When it was defaulting it tried to use absolute "/" locally. Should I revert that back?
Perhaps setting it to "." by default for local and "/" for cloud-based would be better?

File: autogpts/autogpt/autogpt/file_storage/gcs.py (hunk 5 above)
Comment (kcze) [reply autogpts/autogpt/autogpt/file_storage/gcs.py]: Should I mark `WORKSPACE_STORAGE_*` as depracated and keep it together with `FILE_STORAGE_*` or just use `WORKSPACE_STORAGE_*`?

File: autogpts/autogpt/autogpt/agents/utils/file_manager.py
Comment (kcze) [reply autogpts/autogpt/autogpt/agents/utils/file_manager.py:1]: I agree, I was hesitant when introducing `FileManager` because of that code duplication but that was the best I could come up with when I started working on this.

I'll make changes.

File: autogpts/autogpt/autogpt/agents/features/agent_file_manager.py (hunk 10 above)
Comment (kcze) [reply autogpts/autogpt/autogpt/agents/features/agent_file_manager.py]: I'll make naming clearer.
I'll remove file logs from agent entirely in another PR, there are other issues I want to address (handling repeated commands well).

File: autogpts/autogpt/autogpt/file_storage/gcs.py (hunk 5 above)
Comment (Pwuts) [reply autogpts/autogpt/autogpt/file_storage/gcs.py]: I propose to just drop `WORKSPACE_*` and make this very clear in the next release. It hasn't been around for long and it isn't a feature that is used by default.

File: autogpts/autogpt/autogpt/app/main.py (hunk 9 above)
Comment (kcze) [reply autogpts/autogpt/autogpt/app/main.py]: Change agent ID and rename folder. It makes sense to rename it since the assigned name may not be something that user wants.

File: autogpts/autogpt/autogpt/app/main.py (hunk 11 above)
Comment (kcze) [reply autogpts/autogpt/autogpt/app/main.py]: This will fix it
> Change agent ID and rename folder.

File: autogpts/autogpt/tests/conftest.py [hunk 12]
```diff
@@ -40,20 +42,12 @@ def app_data_dir(tmp_project_root: Path) -> Path:
...
-def workspace(workspace_root: Path) -> FileWorkspace:
-    workspace = LocalFileWorkspace(FileWorkspaceConfiguration(root=workspace_root))
-    workspace.initialize()
-    return workspace
+def storage(app_data_dir: Path) -> FileStorage:
+    storage = LocalFileStorage(
+        FileStorageConfiguration(root=app_data_dir, restrict_to_root=False)
+    )
+    storage.initialize()
+    return storage
```
Comment (Pwuts) [inline autogpts/autogpt/tests/conftest.py:50]: image

File: autogpts/autogpt/tests/unit/test_gcs_file_storage.py
Comment (Pwuts) [inline autogpts/autogpt/tests/unit/test_gcs_file_storage.py:1]: image
```
=========================================================================================== FAILURES ===========================================================================================
_______________________________________________________________________________________ test_rename_file _______________________________________________________________________________________

gcs_storage_with_files = GCSFileStorage(bucket='test-bucket-f25888e5', root=/workspaces/AutoGPT-some-unique-task-id)

    def test_rename_file(gcs_storage_with_files: GCSFileStorage):
        for file_name, _ in TEST_FILES:
            new_name = str(file_name) + "_renamed"
            gcs_storage_with_files.rename(file_name, new_name)
>           assert gcs_storage_with_files.exists(new_name)
E           AssertionError: assert False
E            +  where False = <bound method GCSFileStorage.exists of GCSFileStorage(bucket='test-bucket-f25888e5', root=/workspaces/AutoGPT-some-unique-task-id)>('existing_test_file_1_renamed')
E            +    where <bound method GCSFileStorage.exists of GCSFileStorage(bucket='test-bucket-f25888e5', root=/workspaces/AutoGPT-some-unique-task-id)> = GCSFileStorage(bucket='test-bucket-f25888e5', root=/workspaces/AutoGPT-some-unique-task-id).exists

tests/unit/test_gcs_file_storage.py:164: AssertionError
_______________________________________________________________________________________ test_rename_dir ________________________________________________________________________________________

gcs_storage_with_files = GCSFileStorage(bucket='test-bucket-f25888e5', root=/workspaces/AutoGPT-some-unique-task-id)

    def test_rename_dir(gcs_storage_with_files: GCSFileStorage):
>       gcs_storage_with_files.rename(NESTED_DIR, "existing/test/dir_renamed")

tests/unit/test_gcs_file_storage.py:169: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = GCSFileStorage(bucket='test-bucket-f25888e5', root=/workspaces/AutoGPT-some-unique-task-id), old_path = PosixPath('workspaces/AutoGPT-some-unique-task-id/existing/test/dir')
new_path = PosixPath('workspaces/AutoGPT-some-unique-task-id/existing/test/dir_renamed')

    def rename(self, old_path: str | Path, new_path: str | Path) -> None:
        """Rename a file or folder in the storage."""
        old_path = self.get_path(old_path)
        new_path = self.get_path(new_path)
        for blob in self._bucket.list_blobs(prefix=f"{old_path}/"):
>           new_name = str(new_path / blob.namez)
E           AttributeError: 'Blob' object has no attribute 'namez'

autogpt/file_storage/gcs.py:174: AttributeError
```

File: autogpts/autogpt/autogpt/file_storage/gcs.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+        """Create a directory in the storage if doesn't exist."""
+        # GCS does not have directories, so we don't need to do anything
+        pass
+
+    def rename(self, old_path: str | Path, new_path: str | Path) -> None:
+        """Rename a file or folder in the storage."""
+        old_path = self.get_path(old_path)
+        new_path = self.get_path(new_path)
+        for blob in self._bucket.list_blobs(prefix=f"{old_path}/"):
+            new_name = str(new_path / blob.namez)
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/file_storage/gcs.py]: `blob.namez` -> typo?

File: autogpts/autogpt/autogpt/file_storage/gcs.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+    def make_dir(self, path: str | Path) -> None:
+        """Create a directory in the storage if doesn't exist."""
+        # GCS does not have directories, so we don't need to do anything
+        pass
+
+    def rename(self, old_path: str | Path, new_path: str | Path) -> None:
+        """Rename a file or folder in the storage."""
+        old_path = self.get_path(old_path)
+        new_path = self.get_path(new_path)
+        for blob in self._bucket.list_blobs(prefix=f"{old_path}/"):
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/file_storage/gcs.py:179]: This should work for folders, but doesn't seem to work for individual files/blobs. `some/file` won't show up in `list_blobs(prefix="some/file/")`.

File: autogpts/autogpt/autogpt/agents/agent.py [hunk 15]
```diff
@@ -139,11 +142,15 @@ def build_prompt(
...
-                    else ""
+                    else (
+                        " Budget very nearly exceeded! Shut down gracefully!\n\n"
+                        if remaining_budget < 0.005
+                        else (
+                            " Budget nearly exceeded. Finish up.\n\n"
+                            if remaining_budget < 0.01
+                            else ""
+                        )
+                    )
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/agents/agent.py]: Just a sidenote that unrelated formatting is generally not desirable in PRs and we wouldn't allow this from external contributors

File: autogpts/autogpt/autogpt/agents/base.py [hunk 16]
```diff
@@ -250,12 +215,14 @@ async def propose_action(self) -> ThoughtProcessOutput:
...
-            if self.config.use_functions_api
-            else [],
+            functions=(
+                get_openai_command_specs(
+                    self.command_registry.list_available_commands(self)
+                )
+                + list(self._prompt_scratchpad.commands.values())
+                if self.config.use_functions_api
+                else []
+            ),
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/agents/base.py]: Same note about unrelated formatting

File: autogpts/autogpt/autogpt/app/agent_protocol_server.py [hunk 17]
```diff
@@ -387,7 +384,7 @@ async def list_artifacts(
         return TaskArtifactsListResponse(artifacts=artifacts, pagination=pagination)
 
     async def create_artifact(
-        self, task_id: str, file: UploadFile, relative_path: str
+        self, agent: Agent, task_id: str, file: UploadFile, relative_path: str
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/app/agent_protocol_server.py]: I have a strong suspicion that this won't work. Have you tested this? Same for `get_artifact`.

**Review by Pwuts — CHANGES_REQUESTED** [review]

File: autogpts/autogpt/autogpt/app/agent_protocol_server.py (hunk 17 above)
Comment (Pwuts) [reply autogpts/autogpt/autogpt/app/agent_protocol_server.py]: The extra parameter causes an error in the API:
```
[2024-03-07 11:52:20,215] [forge.sdk.routes.agent_protocol] [ERROR]     ❌  Error whilst trying to upload artifact: 4da651b3-8eae-4d5c-97f8-89bd3e7ae8fd
Traceback (most recent call last):
  File "/home/reinier/code/agpt/Auto-GPT/autogpts/autogpt/.venv/lib/python3.11/site-packages/forge/sdk/routes/agent_protocol.py", line 532, in upload_agent_task_artifacts
    artifact = await agent.create_artifact(task_id, file, relative_path)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: AgentProtocolServer.create_artifact() missing 1 required positional argument: 'relative_path'
```

`task_id`, `file` and `relative_path` are parameters of the API endpoint. `agent` is not, but is an internal entity linked to the task (`task_id`). That's why we need `_get_task_agent_file_workspace`.

File: autogpts/autogpt/autogpt/agents/agent.py (hunk 15 above)
Comment (kcze) [reply autogpts/autogpt/autogpt/agents/agent.py]: That wasn't intentional; black extension auto-format in vs code.

File: autogpts/autogpt/autogpt/app/agent_protocol_server.py (hunk 17 above)
Comment (kcze) [reply autogpts/autogpt/autogpt/app/agent_protocol_server.py]: I brought `_get_task_agent_file_workspace` back

**Review by Pwuts — APPROVED** [review]
Changes look good and tests pass. LGTM!

File: autogpts/autogpt/autogpt/agents/features/agent_file_manager.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+    async def save_state(self) -> None:
+        """Save the agent's state to the state file."""
+        state: BaseAgentSettings = getattr(self, "state")
+        await self.files.write_file(self.files.root / self.STATE_FILE, state.json())
+
+    def change_agent_id(self, new_id: str):
+        """Change the agent's ID and update the file storage accordingly."""
+        state: BaseAgentSettings = getattr(self, "state")
+        # Rename the agent's files and workspace
+        self._file_storage.rename(f"agents/{state.agent_id}", f"agents/{new_id}")
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/agents/features/agent_file_manager.py:80]: "Save as" is supposed to behave like any "Save as" function and save the agent in a new location (under a new ID), not rename the current storage location. Could you fix this in a follow-up PR?

File: autogpts/autogpt/test_file.txt
Comment (Pwuts) [inline autogpts/autogpt/test_file.txt:1]: yeet

Comment (kcze) [reply autogpts/autogpt/test_file.txt:1]: sneaky

File: autogpts/autogpt/autogpt/agents/features/agent_file_manager.py (hunk 18 above)
Comment (kcze) [reply autogpts/autogpt/autogpt/agents/features/agent_file_manager.py:80]: Will be fixed

**Comment by Pwuts** [comment]
Fails to deploy to GCP:
```
ERROR 2024-03-12T14:47:31.371921Z Traceback (most recent call last): File "<string>", line 1, in <module> File "/venv/agpt-9TtSrW0h-py3.10/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
DEFAULT 2024-03-12T14:47:31.371932Z return self.main(*args, **kwargs)
DEFAULT 2024-03-12T14:47:31.371941Z File "/venv/agpt-9TtSrW0h-py3.10/lib/python3.10/site-packages/click/core.py", line 1078, in main
DEFAULT 2024-03-12T14:47:31.371945Z rv = self.invoke(ctx)
DEFAULT 2024-03-12T14:47:31.371951Z File "/venv/agpt-9TtSrW0h-py3.10/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
DEFAULT 2024-03-12T14:47:31.371955Z return _process_result(sub_ctx.command.invoke(sub_ctx))
DEFAULT 2024-03-12T14:47:31.371959Z File "/venv/agpt-9TtSrW0h-py3.10/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
DEFAULT 2024-03-12T14:47:31.371963Z return ctx.invoke(self.callback, **ctx.params)
DEFAULT 2024-03-12T14:47:31.371968Z File "/venv/agpt-9TtSrW0h-py3.10/lib/python3.10/site-packages/click/core.py", line 783, in invoke
DEFAULT 2024-03-12T14:47:31.371972Z return __callback(*args, **kwargs)
DEFAULT 2024-03-12T14:47:31.371976Z File "/app/autogpt/app/cli.py", line 272, in serve
DEFAULT 2024-03-12T14:47:31.371980Z run_auto_gpt_server(
DEFAULT 2024-03-12T14:47:31.371984Z File "/app/autogpt/core/runner/client_lib/utils.py", line 60, in wrapper
DEFAULT 2024-03-12T14:47:31.371988Z return asyncio.run(f(*args, **kwargs))
DEFAULT 2024-03-12T14:47:31.371992Z File "/usr/local/lib/python3.10/asyncio/runners.py", line 44, in run
DEFAULT 2024-03-12T14:47:31.371995Z return loop.run_until_complete(main)
DEFAULT 2024-03-12T14:47:31.372Z File "/usr/local/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
DEFAULT 2024-03-12T14:47:31.372003Z return future.result()
DEFAULT 2024-03-12T14:47:31.372007Z File "/app/autogpt/app/main.py", line 351, in run_auto_gpt_server
DEFAULT 2024-03-12T14:47:31.372011Z file_storage = get_storage(
DEFAULT 2024-03-12T14:47:31.372015Z File "/app/autogpt/file_storage/__init__.py", line 37, in get_storage
DEFAULT 2024-03-12T14:47:31.372019Z return GCSFileStorage(config)
DEFAULT 2024-03-12T14:47:31.372023Z File "/app/autogpt/file_storage/gcs.py", line 36, in __init__
DEFAULT 2024-03-12T14:47:31.372026Z assert self._root.is_absolute()
DEFAULT 2024-03-12T14:47:31.372043Z AssertionError

```
